### ISSUE TITLE
klog should de-duplicate k/v pairs

### ISSUE THREAD

**Issue by sbueringer** [kubernetes/klog#377]
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

I ran the following code:
```go
	log := klog.Background()

	log = log.WithValues("abc", "def")
	log = log.WithValues("abc", "def")
	log = log.WithValues("abc", "def")
	log.Info("test")
```

**What did you expect to happen:**

Should log
```bash
{"ts":1683805678697.2798,"caller":"cluster-api/main.go:230","msg":"test","v":0,"abc":"def"}
```

**Anything else you would like to add:**
Did log
```bash
{"ts":1683805678697.2798,"caller":"cluster-api/main.go:230","msg":"test","v":0,"abc":"def","abc":"def","abc":"def"}
```

[Miscellaneous information that will assist in solving the issue.]
* I'm using component-base/logs to configure the format etc.
* ~ the same happens with the text log format
* I'm using klog `v2.90.1`, but I could also reproduce it with `v2.100.1`

I can also provide a minimal repository if it helps to reproduce.

**Comment by sbueringer** [comment]
cc @pohly

**Comment by pohly** [comment]
So you are using zap, right? zap does not de-duplicate and, as far as I know, never has.

klog doesn't handle WithValues for zap and zapr just [passes it through to zap](https://github.com/go-logr/zapr/blob/6e2be81dcddab620aebd76f9e3dd6095efe9ef75/zapr.go#L220-L224).

**Comment by sbueringer** [comment]
Partially. For the json log format we use zap, for test it's just klog as far as I see.

We have the following setup:

```go
import (
         ...
	"k8s.io/component-base/logs"
	logsv1 "k8s.io/component-base/logs/api/v1"
	_ "k8s.io/component-base/logs/json/register"
	"k8s.io/klog/v2"
        ...
)

var (
	logOptions                    = logs.NewOptions()
)

func InitFlags(fs *pflag.FlagSet) {
	logsv1.AddFlags(logOptions, fs)
        ...
}

func main() {
	InitFlags(pflag.CommandLine)
	pflag.CommandLine.SetNormalizeFunc(cliflag.WordSepNormalizeFunc)
	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
	pflag.Parse()

	if err := logsv1.ValidateAndApply(logOptions, nil); err != nil {
		setupLog.Error(err, "unable to start manager")
		os.Exit(1)
	}

        // Just added like that to reproduce easily
	log := klog.Background()

	log = log.WithValues("abc", "def")
	log = log.WithValues("abc", "def")
	log = log.WithValues("abc", "def")
	log.Info("test")
}
```
(actual file we use: https://github.com/kubernetes-sigs/cluster-api/blob/main/main.go)

**Comment by pohly** [comment]
I see. De-duplication currently happens between `WithValues` key/value pairs and the parameters passed to the log call:
```
// de-duplicates:
logger.WithValues("abc", "def").Info("hello", "abc", "def")
// doesn't de-duplicate:
logger.WithValues("abc", "def").WithValues("abc", "def").Info("hello")
```

The reason is that de-duplication is more expensive (`MergeKVs` vs. `WithValues`):
https://github.com/kubernetes/klog/blob/6ded8085b27eeaf68a13d583a382b125ef1482ee/internal/serialize/keyvalues.go#L32-L97

My rationale at the time was that it's unlikely that different parts of the code would use `WithValues` with the same key. Paying the price of checking for duplicates in each `WithValues` call then wouldn't be worth it because they never or only rarely occur.

How often does this occur in your experience? Is it worth making `WithValues` slower for it? Finally, remember that this will only help with text output, not with JSON.

**Comment by pohly** [comment]
The other aspect to consider: `WithValues` and thus de-duplication would get executed whether the call chain actually logs anything or not. That's not the case for de-duplication when recording an event (which is done now), because that by definition only happens when really writing the log event.

Perhaps it would be okay to de-duplicate *all* key/value pairs during encoding, but that then also leads to more runtime overhead.

**Comment by sbueringer** [comment]
> How often does this occur in your experience? Is it worth making WithValues slower for it? Finally, remember that this will only help with text output, not with JSON.

In my case the following happens:
* Controller-runtime itself adds certain k/v pairs before calling reconcile logic or webhook logic
* The code implementing the reconcile or webhook might add the same k/v pair again

This can happen, but (at least in Cluster API) not that frequently.

Considering that it would only help with text output I think it's not worth making `WithValues` slower for the happy path.

I just noticed that the k/v pair wasn't deduplicated and I thought I remembered that it was deduplicated a while back.
